File: leetcode/745_Prefix_and_Suffix_Search.py
from typing import List, Dict
from unittest import TestCase

class Node:
    def __init__(self):
        self.chars = [None] * 27
        self.max_idx = None


class WordFilter:
    def __init__(self, words: List[str]):
        self.trie = Node()
        word_dic: Dict[str, int] = {}
        for idx, word in enumerate(words):
            word_dic[word] = idx
        for word, idx in word_dic.items():
            word = "{" + word
            self._append_word(self.trie, word, idx)
            for ch in word[::-1]:
                word = ch + word
                self._append_word(self.trie, word, idx)

    def f(self, prefix: str, suffix: str) -> int:
        word = suffix + "{" + prefix
        return self._search_word(self.trie, word)
                
    def _append_word(self, node: Node, word: str, idx: int):
        a_ord = ord("a")
        for ch in word:
            ch_i = ord(ch) - a_ord
            if node.chars[ch_i] is None:
                node.chars[ch_i] = Node()
            node = node.chars[ch_i]
            node.max_idx = idx

    def _search_word(self, node: Node, prefix: str) -> int:
        a_ord = ord("a")
        for ch in prefix:
            ch_i = ord(ch) - a_ord
            if node.chars[ch_i] is None:
                return -1
            node = node.chars[ch_i]
        return node.max_idx

class WordFilterTest(TestCase):
    def test_1(self):
        word_filter = WordFilter(["apple"])
        self.assertEqual(0, word_filter.f("a", "e"))

    def test_2(self):
        word_filter = WordFilter(["apple", "fruit", "appdddle"])
        self.assertEqual(2, word_filter.f("ap", "le"))

File: leetcode/745_Prefix_and_Suffix_Search.rs
struct Node {
    chars: [Option<Box<Node>>; 27],
    max_idx: i16,
}

impl Node {
    fn new() -> Self {
        Self { chars: Default::default(), max_idx: -1 }
    }
}

struct WordFilter {
    root: Box<Node>,
}

impl WordFilter {

    fn new(words: Vec<String>) -> Self {
        let mut root = Box::new(Node::new());
        for (idx, word) in words.into_iter().enumerate() {
            Self::append_word(&mut root, word.as_bytes(), idx as i16);
        }
        Self { root }
    }
    
    fn f(&self, prefix: String, suffix: String) -> i32 {
        let prefix = prefix.as_bytes();
        let suffix = suffix.as_bytes();
        let mut node = &self.root;
        for ch in [suffix, &[b'{'], prefix].concat() {
            let ch = ch - b'a';
            if node.chars[ch as usize].is_none() {
                return -1;
            }
            node = node.chars[ch as usize].as_ref().unwrap();
        }
        node.max_idx as i32
    }

    fn append_word(root: &mut Box<Node>, word: &[u8], idx: i16) {
        let n = word.len();
        let word = [word, &[b'{'], word].concat();
        for i in 0..=n {
            let mut node = &mut *root;
            for ch in &word[i..] {
                let ch = ch - b'a';
                if node.chars[ch as usize].is_none() {
                    node.chars[ch as usize] = Some(Box::new(Node::new()));
                }
                node = node.chars[ch as usize].as_mut().unwrap();
                node.max_idx = idx;
                    
            }
        }
    }
}

#[test]
fn test_1() {
    let word_filter = WordFilter::new(vec!["apple".to_string()]);
    assert_eq!(0, word_filter.f("a".to_string(), "e".to_string()));
}
